### ISSUE TITLE
Stacking multiple results on the same call configuration instance doesn't work

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#534]
Sorry if the title is not clear, here's an example to illustrate the problem.

I expected the two following snippets to be equivalent:

```c#
// Snippet 1
A.CallTo(() => foo.Get()).Returns(42);
A.CallTo(() => foo.Get()).Returns(100).Twice();

// Snippet 2
var call = A.CallTo(() => foo.Get());
call.Returns(42);
call.Returns(100).Twice();
```

But apparently they're not. With the first snippet, successive calls to `foo.Get()` return 100, 100, 42, 42, 42, 42 ... (as expected). But with the second snippet, successive calls to `foo.Get()` return 100, 100, 0, 0, 0, 0... The first `Returns()` is ignored; the second `Returns()` apparently overwrites the first, rather than stacking upon it.

Is this the expected behavior? If it is, it's quite confusing, and goes against the principle of least surprise. I sometimes store the result of `A.CallTo()` to avoid repeating it multiple times (usually because I want to configure the call AND verify that it happened), and I certainly didn't expect this behavior.

**Comment by blairconrad** [comment]
@thomaslevesque I would've had the same expectation that you did!

**Comment by thomaslevesque** [comment]
Looking at the code, I understand why it works that way. Each `A.CallTo(...)` creates a new rule for the fake, and each rule has only one return value configuration. So, specifying the return value again just overwrites the previous configuration. It kind of makes sense, but it's not very intuitive...

Perhaps it would be possible to extend the API to allow something like this:

    A.CallTo(() => foo.Get())
        .Returns(100).Twice()
        .Then.Returns(42);

(note that the order is reversed to match the "natural" order, rather than the stacking order)

`Then` would actually add a new rule for the fake, based on the same call specification.

**Comment by blairconrad** [comment]
@thomaslevesque, It may be possible. I think it would be a little nicer to adjust things to support your original example, though, if that didn't break everything else…

**Comment by thomaslevesque** [comment]
> I think it would be a little nicer to adjust things to support your original example, though, if that didn't break everything else…

I agree, but I suspect it would be a pretty big undertaking, with high risk of regression... I'm not sure it's worth it.

**Comment by blairconrad** [comment]
I've thought a little more about this.
I'm not sure that the behaviour is actually that bad. It's maybe a bit surprising, but is defensible.
In your original example:

```c#
A.CallTo(() => foo.Get()).Returns(42);
A.CallTo(() => foo.Get()).Returns(100).Twice();
```

We have a `A.CallTo(() => foo.Get())` whose results we manipulate by saying `.Returns(42)`.
Then we have `A.CallTo(() => foo.Get())` again. This returns (it's not unreasonable to assume) a new object, potentially unrelated to the previous one.
I think interpreting this as referring to a different call does make some sense.

Later, we have

```c#
var call = A.CallTo(() => foo.Get());
call.Returns(42);
call.Returns(100).Twice();
```

In this case, we're saving away the first result. This can be interpreted as talking about a given call. Then the code reads as "I want to return 42. No, never mind. I want to return 100 twice.".

With this interpretation, I am backpedalling on my desire to support the original example.

**Comment by thomaslevesque** [comment]
Yes, I agree this interpretation makes sense as well, but it's a little counter-intuitive... Anyway, I don't think it's a big enough problem to justify the large refactoring that would be required to fix it.

**Comment by blairconrad** [comment]
The more I think about this, I'm wondering if the best course isn't to detect and disallow setting conflicting behaviour on a given call configuration instance.

**Comment by thomaslevesque** [comment]
Yes, that's probably the safest way to handle it. I think it would be pretty easy to do. Maybe I'll give it a try in the next few days.

**Comment by adamralph** [comment]
:+1:

**Comment by thomaslevesque** [comment]
@blairconrad @adamralph, I submitted a PR for this change. I handled this in the `BuildableCallRule.Applicator` setter, since apparently all methods that define method behavior end up setting this property. Not sure it was the best way to do it; waiting for your comments.

**Comment by blairconrad** [comment]
Thanks, @thomaslevesque. We appreciate it. I'll make sure to take a look by the weekend's end.

**Comment by thomaslevesque** [comment]
Continuing the discussion from the pull request

> I would've probably added specs so it would be obvious what external behaviour we're testing (the Applicator setting is all well and good, but even to experienced insiders (if I can consider myself such) it doesn't scream "Returns" or "Throws").

Do you mean tests in the FakeItEasy.Specs project? I can add them if needed. I'm not familiar with Xbehave, but there are enough examples to get me going.

> In addition, a number of StyleCop warnings have been introduced, which aren't huge deals, but if you keep an eye out in round two, I would consider it a favour.

Ah, forgot to check the error list... I will fix the warnings.

> My main concern is that this change addresses multiple Returns, Throws, and CallsBaseMethods, but nothing else, and would leave us with inconsistent behaviour, not that we don't have that now.

Why is it a problem? As far as I can tell, they're the only methods that define the behavior for a rule. Well, there's also Invoke, but apparently it allows to specify multiple actions, so it wouldn't be an error to call it multiple times. So the inconsistency is already there, as you mentioned... The goal of this change is to prevent from doing something that wouldn't work as the user might expect.

**Comment by thomaslevesque** [comment]
> Do you mean tests in the FakeItEasy.Specs project? I can add them if needed. I'm not familiar with Xbehave, but there are enough examples to get me going.

Actually I can't find a spec where Returns and Throws are tested... Should I create a new Specs class ?

**Comment by thomaslevesque** [comment]
I added a few scenarios in ConfigurationSpecs. Not sure if it's the right place to put them, or if they're correctly implemented...

**Comment by blairconrad** [comment]
@thomaslevesque, I'm back a little.

In no particular order:

Yes, ConfigurationSpecs seems a good place. I'm not a BDD/XBehave expert (yet), but am working on it!

>>My main concern is that this change addresses multiple Returns, Throws, and CallsBaseMethods, but nothing else, and would leave us with inconsistent behaviour, not that we don't have that now.

> Why is it a problem? As far as I can tell, they're the only methods that define the behavior for a rule. Well, there's also Invoke, but apparently it allows to specify multiple actions, so it wouldn't be an error to call it multiple times. So the inconsistency is already there, as you mentioned... The goal of this change is to prevent from doing something that wouldn't work as the user might expect.

As you point out, `Invokes` does allow the user to specify multiple actions.
Which can be explained away, but some users may be confused about why they are allowed to specify multiple `Invokes`es, when everything else is locked down to one invocation.
(I know that it's possible for multiple `Invokes` to not interfere, when multiple `Returns` will interfere with one another, but it may not be immediately apparent to the user.)

If someone writes

```c#
call.Invokes(someMethod);
call.Invokes(someOtherMethod).Once();
```

why wouldn't they expect the first call to run `someOtherMethod` and the second one to run `someMethod`? It could be a big surprise when both methods are run on the first call.

That behaviour could conceivably be put aside or explained away in documentation, but there's still `AssignsOutAndRefParameters`, which as your code is now can be called multiple times, with the last one in winning. This is exactly analogous to how the `Returns` behaviour is in the released FakeItEasy, and I think we're best off cleaning it up while cleaning up `Returns` and its ilk.

**Comment by blairconrad** [comment]
My mistake. `Invokes` doesn't provide access to `Once()`.
I'm constantly learning things about FakeItEasy!

**Comment by thomaslevesque** [comment]
> why wouldn't they expect the first call to run someOtherMethod and the second one to run someMethod? It could be a big surprise when both methods are run on the first call.

Good point. The code explicitly supports specifying multiple actions with `Invokes` (it maintains a list of actions), but IMO it makes no sense (you can put everything in the same action). Ideally I think the ability to do that should be removed, but it would be a breaking change...

> That behaviour could conceivably be put aside or explained away in documentation, but there's still AssignsOutAndRefParameters, which as your code is now can be called multiple times, with the last one in winning. This is exactly analogous to how the Returns behaviour is in the released FakeItEasy, and I think we're best off cleaning it up while cleaning up Returns and its ilk.

Yes, didn't think about that one. I'll see if I can handle `AssignsOutAndRefParameters` as well.

**Comment by blairconrad** [comment]
>  Ideally I think the ability to [specify more than one Invokes] should be removed, but it would be a breaking change...

I sort of feel the same way. I wouldn't really want to remove functionality without a good reason, I'm not sure of the benefit. And if order of operations becomes important (I hope it wouldn't, but people write different code. I once met a fella who tried to use `Returns` on the same configuration object! :stuck_out_tongue_winking_eye: ), then these `Invokes` are about the only thing in FakeItEasy that are run FIFO instead of LIFO, which is just begging for confusion.

The "breaking change" part is troubling, but as we prepare for 2.0, if every we're going to break something, now's the time. We'd want input from other @FakeItEasy/owners, however…

**Comment by thomaslevesque** [comment]
I updated the code to use `Record.Exception`, and prevent multiple calls to `AssignsOutAndRefParameters`. I also updated the exception message. Not sure it's clearer, but at least it doesn't mention the rule anymore.

**Comment by blairconrad** [comment]
This issue has been fixed in [release 2.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.0.0).
